### PR TITLE
Convert LLVM to monorepo build

### PR DESCRIPTION
**Pull request by vimproved**
This has been discussed in `#gentoo-llvm`, and I decided to try my hand at this. This is very much a work in progress, but it should at least be working.

TODO:
- General cleanups
- PGO. PGO with the cmake cache is only possible if compiler-rt is in the tree, which it ideally would not be since it is remaining split. It should be possible to do PGO with external compiler-rt, but it may require some "interesting" stuff. Alternatively, we could just use the PGO cache but not install compiler-rt (probably the better solution)

### CODE REVIEW

File: profiles/desc/llvm_projects.desc [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Copyright 1999-2022 Gentoo Authors.
```
Comment (thesamesam) [inline profiles/desc/llvm_projects.desc:1]:
```suggestion
# Copyright 1999-2023 Gentoo Authors
```

File: profiles/desc/llvm_runtimes.desc [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# Copyright 1999-2022 Gentoo Authors.
```
Comment (thesamesam) [inline profiles/desc/llvm_runtimes.desc:1]:
```suggestion
# Copyright 1999-2023 Gentoo Authors
```

File: sys-devel/clang/clang-18.ebuild [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+EAPI=8
+
+DESCRIPTION="C language family frontend for LLVM (metapackage)"
+HOMEPAGE="https://llvm.org/"
+
+LICENSE="metapackage"
+SLOT="${PV}"
+
+RDEPEND="
+	sys-devel/llvm:${PV}[llvm_projects_clang]
```
Comment (thesamesam) [inline sys-devel/clang/clang-18.ebuild:13]: If `llvm_projects_clang` isn't a USE flag for LLVM, then it's an unmigrated ebuild, so we don't want it (assume it's off):
```suggestion
	sys-devel/llvm:${PV}[llvm_projects_clang(-)]
```

File: sys-devel/lld/lld-18.ebuild [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+DESCRIPTION="The LLVM linker (link editor) (metapackage)"
+HOMEPAGE="https://llvm.org/"
+
+LICENSE="metapackage"
+SLOT="${PV}"
+KEYWORDS=""
+
+RDEPEND="
+	sys-devel/llvm:${PV}[llvm_projects_lld]
```
Comment (thesamesam) [inline sys-devel/lld/lld-18.ebuild:14]:
```suggestion
	sys-devel/llvm:${PV}[llvm_projects_lld(-)]
```

File: sys-devel/llvm/llvm-18.0.0.9999.ebuild [hunk 5]
```diff
@@ -65,18 +80,28 @@ BDEPEND="
...
+	)
 "
 
-LLVM_COMPONENTS=( llvm cmake third-party )
+LLVM_COMPONENTS=( llvm clang clang-tools-extra cmake compiler-rt libunwind/include/mach-o lld runtimes third-party )
 LLVM_MANPAGES=1
 LLVM_USE_TARGETS=provide
 llvm.org_set_globals
 
+pkg_setup() {
```
Comment (thesamesam) [inline sys-devel/llvm/llvm-18.0.0.9999.ebuild:101]: Do we need this explicitly? From a quick skim, python-single-r1 is the only inherited eclass providing pkg_setup so it'll get called anyway

**Review by thesamesam — COMMENTED** [review]
I expect this to be a bit controverisal, but I genuinely think we need to consider it, for a few reasons:
* The upgrade process right now is brittle because of ABI (this is especially a problem between RCs)
* The upgrade process right now is also brittle because of upgrade ordering and concerns with keeping a working Clang throughout (because you need it to build the LLVM other packages)
* We can't do PGO or (effective) LTO without this - the PGO point is obvious, but LTO Isn't as effective with dynamically linked libraries.
* Upstream aren't super interested in the split component build so we end up chasing after it quite a lot
* It means we can stop worrying about `LLVM_TARGETS` breaking ABI because everything becomes internal, so we end up clawing back a bunch of build time
* It should make bootstrapping for LLVM-based systems a bit easier, especially thinking of e.g. macOS
* It makes downgrading/upgrading a lot easier because there's a single ebuild to dig out from git history, rather than several interconnected packages

Considering the opposite, what are the advantages of the current approach? The only thing I can think of right now is saving build time for components you don't need (which is negated by the LLVM_TARGETS point above & you can pick which components you want with this anyway), and a smaller ebuild (which isn't an advantage by itself).

While LLVM does support the split approach more than GCC does, it doesn't seem like it really works that well in reality and is about as fragile as I'd expect when people try to use it as their system toolchain. See also https://bugs.gentoo.org/710378 and https://bugs.gentoo.org/812938.

**Comment by vimproved** [comment]
>     * A smaller ebuild (which isn't an advantage by itself)
I'd argue it's the opposite really -- doing this condenses a lot of the redundant things from clang into llvm. The LLVM ebuild becomes larger, but the other ebuilds are being completely eliminated other than as metapackages. If the LLVM ebuild starts getting too out of hand, parts of it could always be split into an `llvm-toolchain.eclass` or something.

**Comment by mgorny** [comment]
You do realize that this actually makes it impossible for me to test LLVM because I don't have that much free space?

**Comment by mgorny** [comment]
Let me elaborate: LLVM is a hog.

Forcing everyone to build everything at once will further limit people's ability to build it on weaker hardware. Disabling targets won't help much (let me remind you we actually checked that when we decided to force them), also I'm not 100% convinced it's going to help with Rust.

Also remember that you'd have to effectively enable all common components by default. It's always the choice of enabling more of them and wasting people's time if they don't need them, or enabling less and having them rebuild everything over and over again whenever they need to enable one additional tiny component.

I don't even want to start considering the resulting mess of the fact that some components support more architectures than others.

Tests aren't 100% perfect either. Just imagine running all the test suites for a few hours just to have compiler-rt tests hang and end up with no result. Or have them suddenly fail because of lit's tests failing randomly.

Monorepo builds are a great idea for binary distros. Gentoo is not a binary distro.

**Comment by mgorny** [comment]
Oh, and I consider this approach of sending a pull request for a rejected idea deeply disrespectful. You're putting me in the position of "I did all the work, and now you're going to reject it, it's your fault".

**Comment by thesamesam** [comment]
> Oh, and I consider this approach of sending a pull request for a rejected idea deeply disrespectful. You're putting me in the position of "I did all the work, and now you're going to reject it, it's your fault".

That's a misrepresentation and I think you should reconsider this comment. Not only that, we've discussed it many times in #gentoo-llvm and asked your opinion on it (and got nothing).

**Comment by thesamesam** [comment]
> Let me elaborate: LLVM is a hog.
> 
> Forcing everyone to build everything at once will further limit people's ability to build it on weaker hardware. Disabling targets won't help much (let me remind you we actually checked that when we decided to force them), also I'm not 100% convinced it's going to help with Rust.
> 

This addresses one of my points, and I'm not sure about how true it is. We should measure again.

> Also remember that you'd have to effectively enable all common components by default. It's always the choice of enabling more of them and wasting people's time if they don't need them, or enabling less and having them rebuild everything over and over again whenever they need to enable one additional tiny component.

What components are we eliding exactly at the moment that are significant? libcxx is the only one I think you can actually avoid pulling that's of significant size, but we can check.

> 
> I don't even want to start considering the resulting mess of the fact that some components support more architectures than others.
> 

We have a mess right now in profiles and the various p.use.forces and REQUIRED_USE. I don't think it makes much difference there.

> Tests aren't 100% perfect either. Just imagine running all the test suites for a few hours just to have compiler-rt tests hang and end up with no result. Or have them suddenly fail because of lit's tests failing randomly.
> 

We can try something like running w/ `nonfatal` and summarising at the end?

> Monorepo builds are a great idea for binary distros. Gentoo is not a binary distro.

You've not addressed the majority of my points, and _you've_ been rather disrespectful in your tone when everything before your comment was diplomatic and focused on the substance.

**Comment by vimproved** [comment]
> Oh, and I consider this approach of sending a pull request for a rejected idea deeply disrespectful. You're putting me in the position of "I did all the work, and now you're going to reject it, it's your fault".

I'm sorry if this came off as disrespectful. I did not get the impression that this was an outright rejected idea, and we've talked about this in `#gentoo-llvm` a few times since rc3 (which broke ABI in the loongarch backend, prompting this conversation), including me sharing a proof of concept for this earlier. I'm not of the mindset of "I did all the work, and now you're going to reject it, it's your fault," if there's a consensus that this isn't a good idea then I won't be offended in any way. I would have done this for fun even if it was clear this was a rejected idea. I put this PR up as a draft to get comments and review, knowing that there was a good chance it would get rejected on concept, not because I thought it should be immediately merged.

**Comment by vimproved** [comment]
> Tests aren't 100% perfect either. Just imagine running all the test suites for a few hours just to have compiler-rt tests hang and end up with no result. Or have them suddenly fail because of lit's tests failing randomly.

If you ever need help with testing I will gladly offer it.

**Comment by vimproved** [comment]
> You do realize that this actually makes it impossible for me to test LLVM because I don't have that much free space?

I'm only getting a final size of 3.7G for the workdir with all runtimes/projects enabled and the X86 and AArch64 targtes (albeit using LTO).

File: sys-devel/llvm/llvm-18.0.0_pre20230906.ebuild [hunk 6]
```diff
@@ -16,13 +16,34 @@ HOMEPAGE="https://llvm.org/"
...
-LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA BSD public-domain rc"
+LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA BSD MIT public-domain rc"
 SLOT="${LLVM_MAJOR}/${LLVM_SOABI}"
 KEYWORDS=""
 IUSE="
-	+binutils-plugin +debug debuginfod doc exegesis libedit +libffi
-	ncurses test xar xml z3 zstd
+	+binutils-plugin +debug debuginfod doc exegesis ieee-long-double libedit
+	+libffi ncurses +pie +static-analyzer test xar xml z3 zstd
+"
```
Comment (wikkyk) [inline sys-devel/llvm/llvm-18.0.0_pre20230906.ebuild:28]: Missing `extras`: (ditto for .9999)

```
 * Messages for package sys-devel/llvm-18.0.0_pre20230906:

 * FAILED postinst: 1
 * You can find additional opt-viewer utility scripts in:
 *   /usr/lib/llvm/18/share/opt-viewer
 * To use these scripts, you will need Python along with the following
 * packages:
 *   dev-python/pygments (for opt-viewer)
 *   dev-python/pyyaml (for all of them)
 * You can find additional utility scripts in:
 *   /usr/lib/llvm/18/share/clang
 * ERROR: sys-devel/llvm-18.0.0_pre20230906::llvm-monorepo failed (postinst phase):
 * 
 * Call stack:
 *          ebuild.sh, line  136:  Called pkg_postinst
 *        environment, line 3711:  Called use 'extra'
 *   phase-helpers.sh, line  256:  Called die
 * The specific snippet of code:
 *   				die "USE Flag '${u}' not in IUSE for ${CATEGORY}/${PF}"
 * 
 * If you need support, post the output of `emerge --info '=sys-devel/llvm-18.0.0_pre20230906::llvm-monorepo'`,
 * the complete build log and the output of `emerge -pqv '=sys-devel/llvm-18.0.0_pre20230906::llvm-monorepo'`.
 * The complete build log is located at '/mnt/scratch/portage/sys-devel/llvm-18.0.0_pre20230906/temp/build.log'.
 * The ebuild environment file is located at '/mnt/scratch/portage/sys-devel/llvm-18.0.0_pre20230906/temp/environment'.
 * Working directory: '/mnt/scratch/portage/sys-devel/llvm-18.0.0_pre20230906/empty'
 * S: '/mnt/scratch/portage/sys-devel/llvm-18.0.0_pre20230906/work/llvm'
```

**Comment by wikkyk** [comment]
I have installed this (18.0.0_pre20230906) and it compiles fine (llvm-16, musl-1.2.4). Timings-wise it doesn't look bad:

    >>> sys-devel/llvm: 2:41:50

vs the current ::gentoo split build:

    >>> sys-devel/llvm: 1:15:40
    >>> sys-devel/clang: 1′04″

(not a proper benchmark - I've had other things running, so the best we can say is it's not *slower*.)

It does produce a working compiler, although it cannot find crt on its own:

```
$ clang-18 
ld.lld: error: cannot open crtbeginS.o: No such file or directory
ld.lld: error: cannot open /lib/llvm/18/bin/../../../../lib/clang/18/lib/linux/libclang_rt.builtins-x86_64.a: No such file or directory
ld.lld: error: cannot open /lib/llvm/18/bin/../../../../lib/clang/18/lib/linux/libclang_rt.builtins-x86_64.a: No such file or directory
ld.lld: error: cannot open crtendS.o: No such file or directory
clang-18: error: linker command failed with exit code 1 (use -v to see invocation)
```

The crt files have been installed to `/lib/llvm/18/lib/clang/18/lib/{x86_64,i386}-linux-musl` as `clang_rt.crtbegin.o`  `clang_rt.crtend.o`	`libclang_rt.builtins.a`. (The i386 crt shouldn't have also been installed, but they're real 32-bit libs.)

However, to make crossdev work, compiler-rt will probably have to remain split, as the crt libs are built for a *tuple* not just an arch and this only installs it for the host tuple.
Another way would be to hollow-out the compiler-rt ebuild and only keep the crossdev bits that @alfredfo and I have figured out so that host crt is installed together with llvm and the crts for additional tuples can be installed separately. I'm not sure if that's worth the maintenance effort, though.

From #gentoo-llvm, so it doesn't get lost:

```
<catcream> vimproved >i'm wondering if compiler-rt will just be built for all the LLVM_TARGETS or what
           no, it needs to target a triple, not architecture                                          
           hence you need to unpack libc headers then compile compiler-rt                             
   <wikky> so that's one of the things that should remain split                                       
<catcream> for the host system there should be no problem doing a monobuild with it                   
           but for cross yeah                                                                         
           same with libc++,c++abi,unwind                                                             
```

**Comment by vimproved** [comment]
> I have installed this (18.0.0_pre20230906) and it compiles fine (llvm-16, musl-1.2.4). Timings-wise it doesn't look bad:
> 
> ```
> >>> sys-devel/llvm: 2:41:50
> ```
> 
> vs the current ::gentoo split build:
> 
> ```
> >>> sys-devel/llvm: 1:15:40
> >>> sys-devel/clang: 1′04″
> ```
> 
> (not a proper benchmark - I've had other things running, so the best we can say is it's similar.)
> 
> It does produce a working compiler, although it cannot find crt on its own:
> 
> ```
> $ clang-18 
> ld.lld: error: cannot open crtbeginS.o: No such file or directory
> ld.lld: error: cannot open /lib/llvm/18/bin/../../../../lib/clang/18/lib/linux/libclang_rt.builtins-x86_64.a: No such file or directory
> ld.lld: error: cannot open /lib/llvm/18/bin/../../../../lib/clang/18/lib/linux/libclang_rt.builtins-x86_64.a: No such file or directory
> ld.lld: error: cannot open crtendS.o: No such file or directory
> clang-18: error: linker command failed with exit code 1 (use -v to see invocation)
> ```
> 
> The crt files have been installed to `/lib/llvm/18/lib/clang/18/lib/{x86_64,i386}-linux-musl` as `clang_rt.crtbegin.o` `clang_rt.crtend.o` `libclang_rt.builtins.a`. (The i386 crt shouldn't have also been installed, but they're real 32-bit libs.)
> 
> However, to make crossdev work, compiler-rt will probably have to remain split, as the crt libs are built for a _tuple_ not just an arch and this only installs it for the host tuple. Another way would be to hollow-out the compiler-rt ebuild and only keep the crossdev bits that @alfredfo and I have figured out so that host crt is installed together with llvm and the crts for additional tuples can be installed separately. I'm not sure if that's worth the maintenance effort, though.
> 
> It does work with sys-libs/compiler-rt-18.0.0_pre20230906::gentoo as well.
> 
> From #gentoo-llvm, so it doesn't get lost:
> 
> ```
> <catcream> vimproved >i'm wondering if compiler-rt will just be built for all the LLVM_TARGETS or what
>            no, it needs to target a triple, not architecture                                          
>            hence you need to unpack libc headers then compile compiler-rt                             
>    <wikky> so that's one of the things that should remain split                                       
> <catcream> for the host system there should be no problem doing a monobuild with it                   
>            but for cross yeah                                                                         
>            same with libc++,c++abi,unwind                                                             
> ```

Yep, i'm aware compiler-rt is broken at the moment. My laptop broke last night while I was trying to fix it.
however, the problem is that if we ever want PGO to work, compiler-rt has to be part of the monorepo build. Perhaps there's a different way to get it to build compiler-rt for all LLVM_TARGETS.

File: sys-devel/llvm/llvm-18.0.0_pre20230906.ebuild (hunk 6 above)
Comment (vimproved) [reply sys-devel/llvm/llvm-18.0.0_pre20230906.ebuild:28]: > Missing `extras`: (ditto for .9999)
> 
> ```
>  * Messages for package sys-devel/llvm-18.0.0_pre20230906:
> 
>  * FAILED postinst: 1
>  * You can find additional opt-viewer utility scripts in:
>  *   /usr/lib/llvm/18/share/opt-viewer
>  * To use these scripts, you will need Python along with the following
>  * packages:
>  *   dev-python/pygments (for opt-viewer)
>  *   dev-python/pyyaml (for all of them)
>  * You can find additional utility scripts in:
>  *   /usr/lib/llvm/18/share/clang
>  * ERROR: sys-devel/llvm-18.0.0_pre20230906::llvm-monorepo failed (postinst phase):
>  * 
>  * Call stack:
>  *          ebuild.sh, line  136:  Called pkg_postinst
>  *        environment, line 3711:  Called use 'extra'
>  *   phase-helpers.sh, line  256:  Called die
>  * The specific snippet of code:
>  *   				die "USE Flag '${u}' not in IUSE for ${CATEGORY}/${PF}"
>  * 
>  * If you need support, post the output of `emerge --info '=sys-devel/llvm-18.0.0_pre20230906::llvm-monorepo'`,
>  * the complete build log and the output of `emerge -pqv '=sys-devel/llvm-18.0.0_pre20230906::llvm-monorepo'`.
>  * The complete build log is located at '/mnt/scratch/portage/sys-devel/llvm-18.0.0_pre20230906/temp/build.log'.
>  * The ebuild environment file is located at '/mnt/scratch/portage/sys-devel/llvm-18.0.0_pre20230906/temp/environment'.
>  * Working directory: '/mnt/scratch/portage/sys-devel/llvm-18.0.0_pre20230906/empty'
>  * S: '/mnt/scratch/portage/sys-devel/llvm-18.0.0_pre20230906/work/llvm'
> ```

ah, this is due to me copy-pasting pkg_postinst from clang without changing the use. I'll fix asap.

**Comment by vimproved** [comment]
Removed compiler-rt. It should probably stay split, I'm sure there's a way to do PGO without installing compiler-rt as well.

**Comment by vimproved** [comment]
Fixing polly at the moment, a patch will be required until https://github.com/llvm/llvm-project/pull/66598 is merged.

**Comment by vimproved** [comment]
@thesamesam @mgorny I would really appreciate some clarity on this. Does this have any possibility of getting merged, or should I just close this?

**Comment by vimproved** [comment]
Since I haven't received a response on this, I'm going to close this. I'll continue the LLVM monorepo work in https://codeberg.org/vimproved/llvm-monorepo-overlay, and I'll create a PR if I get it to a place where I think it's ready.

**Comment by thesamesam** [comment]
Sorry, I hadn't seen your previous comment. That sounds like it's for the best for now, even if I think this is an inevitable approach we'll have to take.

**Pull request closed without merging**